### PR TITLE
Support integral type range window

### PR DESCRIPTION
**Pull request by wbo4958**
This PR tries to add integral type (Boolean exclusive) support for range window frame. With this PR, Rapids can support below types for range window

| | |
|---|---|
|order by column | ByteType, ShortType, IntegerType, LongType, Date, TimestampType |
|boundary | ByteType, ShortType, IntegerType, LongType, IntegerType, CalendarIntervalType |

This PR depends on https://github.com/rapidsai/cudf/pull/7909

### CODE REVIEW

**Comment by sameerz** [comment]
This should target release 0.6, since this depends on https://github.com/rapidsai/cudf/pull/7909 which is targeting RAPIDS 0.20.

**Comment by gerashegalov** [comment]
Closing in favor of the smaller sub-PRs:
- #2222
- #2206
- #2192
- #2210
- #2172
- #2168

**Comment by jlowe** [comment]
This PR has not seen any activity for quite some time.  Is it still relevant?  As @sameerz mentioned earlier, branch-0.5 is closed for everything except critical fixes, so this PR needs to be retargeted to branch-0.6 if it is still needed.

**Comment by sameerz** [comment]
This needs to be retargeted to branch-0.6 now that https://github.com/rapidsai/cudf/pull/7866 (the dependency of https://github.com/rapidsai/cudf/pull/7909) has been merged.

**Comment by wbo4958** [comment]
Yes, I will retarget this today.

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
@revans2 @jlowe @mythrocks Could you help to review this PR. BTW, this PR depends on https://github.com/rapidsai/cudf/pull/7909. Thx

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 1]
```diff
@@ -481,6 +481,36 @@ public static DType getNonNestedRapidsType(DataType type) {
     return result;
   }
 
+  /**
+   * Create a Scalar for boundary value
+   * @param type the type of order by column
+   * @param value boundary value
+   * @return Scalar holding boundary value
+   */
+  public static Scalar createRangeWindowBoundary(DType type, long value) {
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 2]
```diff
@@ -168,14 +208,18 @@ case class GpuWindowExpression(windowFunction: Expression, windowSpec: GpuWindow
     frameType match {
       case RowFrame   => evaluateRowBasedWindowExpression(cb)
       case RangeFrame =>
-        val lower = getRangeBasedLower(windowFrameSpec)
-        val upper = getRangeBasedUpper(windowFrameSpec)
-        if (lower == Int.MinValue && upper == Int.MaxValue) {
+        val orderByType = cb.column(windowSpec.partitionSpec.length).dataType()
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala]: Done, Added some comments

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 3]
```diff
@@ -229,15 +277,31 @@ case class GpuWindowExpression(windowFunction: Expression, windowSpec: GpuWindow
...
-              windowSpec.orderSpec,
-              numGroupingColumns))
-        withResource(table
-            .groupBy(0 until numGroupingColumns: _*)
-            .aggregateWindowsOverTimeRanges(agg)) { aggResultTable =>
-          aggResultTable.getColumn(0).incRefCount()
+          .slice(totalExtraColumns, boundRangeProjectList.length)
+
+        // get the preceding/following scalar to construct WindowOptions
+        val orderByType = table.getColumn(numGroupingColumns).getType
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:265]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 4]
```diff
@@ -61,6 +60,44 @@ class GpuWindowExpressionMeta(
       -1
   }
 
+  /**
+   * Check if the boundary is unbounded
+   * @param boundary boundary expression
+   * @return Boolean
+   */
+  private def rangeBoundaryIsUnBounded(boundary: Expression): Boolean = boundary match {
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:68]: Done

File: tests/src/test/scala/com/nvidia/spark/rapids/WindowFunctionSuite.scala [hunk 5]
```diff
@@ -336,6 +336,72 @@ class WindowFunctionSuite extends SparkQueryCompareTestSuite {
     testAllWindowAggregations(windowClause)
   }
 
+  testSparkResultsAreEqual("[Window] [RANGE] [ASC] [Integral Type]",
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply tests/src/test/scala/com/nvidia/spark/rapids/WindowFunctionSuite.scala:339]: yeah, make sense

**Comment by revans2** [comment]
@wbo4958 if I understand this correctly, Spark will overflow when doing the value calculation and get the wrong answer? But cudf is not overflowing so the answer is technically "wrong". What does cudf do in the case of a long value that might overflow? Is it just treating everything as longs internally?

**Comment by revans2** [comment]
@mythrocks @jlowe @tgravescs @sameerz do either of you have alternative ideas on how we should handle this?

**Comment by wbo4958** [comment]
Thx @revans2 @jlowe @mythrocks I updated this PR to follow Bobby's comments. Please help to review again. Thx very much

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 6]
```diff
@@ -808,6 +808,68 @@ object RapidsConf {
       .booleanConf
       .createWithDefault(false)
 
+  private def getRangeWindowTypeDesc(dt: String, lower: String, upper: String) =
+    s"When order by column is $dt type and the range boundary calculated for a row has overflow, " +
+      s"Spark CPU will get different result comparing GPU.<br>For example, let's assume we have " +
+      s"two columns, one partition column named `uid` and the other column named `dollars`. " +
+      s"And the data is like below,<br>" +
+      s"<code>" +
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:816]: I don't think a config description is an appropriate place for an involved example with a sample table.  Check out both the generated .md file and how it looks in Github, it's kind of a mess in both cases.  IMO this should just state the possibility for overflow.  Any detailed discussion like the example here should be put in the windowing section in `docs/compatibility.md` which should be updated as part of this PR.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 7]
```diff
@@ -808,6 +808,68 @@ object RapidsConf {
...
+    .doc(getRangeWindowTypeDesc("short", "32767", "32767").replace("|", "\\|"))
+    .booleanConf
+    .createWithDefault(false)
+
+  val ENABLE_RANGE_WINDOW_INT = conf("spark.rapids.sql.window.range.int.enabled")
+    .doc(getRangeWindowTypeDesc("int", "2147483647", "2147483647").replace("|", "\\|"))
+    .booleanConf
+    .createWithDefault(true)
+
+  val ENABLE_RANGE_WINDOW_LONG = conf("spark.rapids.sql.window.range.long.enabled")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:867]: Curious, do we really want/need this config?  In what case would a user disable `long` ranges but not others or windowing entirely?  I'm not saying it must be removed, just curious if others think it's better to have the config for type completeness despite not having a use-case for it.  We have a lot of configs, so maybe I'm a bit sensitive to adding any more than necessary.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 8]
```diff
@@ -1096,6 +1158,8 @@ object RapidsConf {
     .booleanConf
     .createWithDefault(true)
 
+
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1161]: Nit: Please avoid unrelated whitespace changes, as it creates a larger patch footprint and more chance of merge conflicts.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 9]
```diff
@@ -94,25 +93,63 @@ class GpuWindowExpressionMeta(
...
+                    if (!conf.isRangeWindowShortEnabled) willNotWorkOnGpu(s"Range window frame" +
+                      s" is not 100% compatible when the order by type is short. To enable it" +
+                      " please set ${RapidsConf.ENABLE_RANGE_WINDOW_SHORT} to true.")
+                  case IntegerType =>
+                    if (!conf.isRangeWindowIntEnabled) willNotWorkOnGpu(
+                      s"Range window frame is not 100% compatible when the order by type is int." +
+                        s" To enable it please set ${RapidsConf.ENABLE_RANGE_WINDOW_INT} to true.")
+                  case LongType =>
+                    if (!conf.isRangeWindowLongEnabled) willNotWorkOnGpu(
+                      s"Range window frame is not 100% compatible when the order by type is long." +
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:134]: This isn't really a true statement, correct?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 10]
```diff
@@ -362,6 +401,68 @@ object GpuWindowExpression {
...
+      case DType.INT32 => Scalar.fromInt(value.toInt)
+      case DType.INT64 => Scalar.fromLong(value)
+      case DType.TIMESTAMP_DAYS => Scalar.durationFromLong(DType.DURATION_DAYS, value)
+      case DType.TIMESTAMP_SECONDS => Scalar.durationFromLong(DType.DURATION_SECONDS, value)
+      case DType.TIMESTAMP_MILLISECONDS =>
+        Scalar.durationFromLong(DType.DURATION_MILLISECONDS, value)
+      case DType.TIMESTAMP_MICROSECONDS =>
+        Scalar.durationFromLong(DType.DURATION_MICROSECONDS, value)
+      case DType.TIMESTAMP_NANOSECONDS => Scalar.durationFromLong(DType.DURATION_NANOSECONDS, value)
+      case _ => Scalar.fromNull(orderByType)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:424]: Do we really want to ignore `value` here and assume the caller wants a null Scalar here, or is throwing an exception more appropriate?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 7 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:867]: Hmm, CPU and GPU will get different results even for long when involving overflow, so better to keep this config just in case the user would like to disable acceleration.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 8 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1161]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala (hunk 9 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:134]: it's true, when overflow even for long, the CPU and GPU will get different results

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala (hunk 10 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:424]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 6 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:816]: Done.

File: docs/compatibility.md [hunk 11]
```diff
@@ -271,13 +271,66 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+### Window Functions
 
 Because of ordering differences between the CPU and the GPU window functions especially row based
 window functions like `row_number`, `lead`, and `lag` can produce different results if the ordering
 includes both `-0.0` and `0.0`, or if the ordering is ambiguous. Spark can produce different results
 from one run to another if the ordering is ambiguous on a window function too.
 
+### Range Window
+
+When the order-by column of a range based window is numeric type like `byte/short/int/long` and the range boundary calculated for a value has overflow, CPU and GPU will get different results.
```
Comment (jlowe) [inline docs/compatibility.md:285]: Please wrap lines to approx. 100 characters to match the style elsewhere in this document.  It makes it easier to read this file from a terminal and also easier to comment on lines for review.

File: docs/compatibility.md [hunk 12]
```diff
@@ -271,13 +271,66 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+
+For example, let's assume we have two columns, one partition-by column named `uid` and the other order-by column named `dollars`. And the dataset is as below.
```
Comment (jlowe) [inline docs/compatibility.md:287]: IMO we don't need to elaborate on partition ID and order-by, that's already covered in the SQL statement below.
```suggestion
For example, consider the following dataset:
```

File: docs/compatibility.md [hunk 13]
```diff
@@ -271,13 +271,66 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+|    1 |      78 |
+|    1 |      84 |
+|    1 |      85 |
+|    1 |      86 |
+|    1 |      92 |
+|    1 |      98 |
++------+---------+
+```
+
+After executing below SQL statement,
```
Comment (jlowe) [inline docs/compatibility.md:316]:
```suggestion
After executing the SQL statement:
```

File: docs/compatibility.md [hunk 14]
```diff
@@ -271,13 +271,66 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+
+``` sql
+SELECT
+ COUNT(dollars) over
+    (PARTITION BY id
+    ORDER BY CAST (dollars AS Byte) ASC
+    RANGE BETWEEN 127 PRECEDING AND 127 FOLLOWING)
+From table
```
Comment (jlowe) [inline docs/compatibility.md:324]:
```suggestion
FROM table
```

File: docs/compatibility.md [hunk 15]
```diff
@@ -271,13 +271,66 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+```
+
+We can get below different results for both CPU and GPU.
```
Comment (jlowe) [inline docs/compatibility.md:327]:
```suggestion
The results will differ between the CPU and GPU due to overflow handling.
```

File: docs/compatibility.md [hunk 16]
```diff
@@ -271,13 +271,66 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+
+``` console
+CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0])
+GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
+```
+
```
Comment (jlowe) [inline docs/compatibility.md:333]: This section should mention the configs added in this PR and/or link to the configs document so users reading this are aware that they can force the CPU behavior if desired.

File: docs/compatibility.md (hunk 12 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md (hunk 11 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done.

File: docs/compatibility.md (hunk 13 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md (hunk 14 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md (hunk 15 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md (hunk 16 above)
Comment (wbo4958) [reply docs/compatibility.md:334]: Done

File: docs/compatibility.md [hunk 17]
```diff
@@ -271,13 +271,75 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+The results will differ between the CPU and GPU due to overflow handling.
+
+``` console
+CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0])
+GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
+```
+
+To disable this operation on the GPU, set
```
Comment (jlowe) [inline docs/compatibility.md:335]: "this operation" is a little too generic, best to be explicit here with something like, "To disable byte-range windowing on the GPU, set"

File: docs/compatibility.md [hunk 18]
```diff
@@ -271,13 +271,75 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+[`spark.rapids.sql.window.range.byte.enabled`](configs.md#sql.window.range.byte.enabled) to true.
+
+We also provide the similar configurations:
```
Comment (jlowe) [inline docs/compatibility.md:338]:
```suggestion
We also provide configurations for other integral range types:
```

File: docs/compatibility.md [hunk 19]
```diff
@@ -271,13 +271,75 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+[`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled),
+[`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled),
+[`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled).
```
Comment (jlowe) [inline docs/compatibility.md:341]: Suggest using bullets here rather than a comma-separated list

**Review by jlowe — COMMENTED** [review]
Minor comments on the docs but otherwise lgtm.  Would be good to hear from @revans2 to make sure he's happy with the rework as well.

File: docs/compatibility.md (hunk 17 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md (hunk 18 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md (hunk 19 above)
Comment (wbo4958) [reply docs/compatibility.md]: Done

File: docs/compatibility.md [hunk 20]
```diff
@@ -332,13 +332,14 @@ CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0
...
 
-We also provide the similar configurations:
-[`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled),
-[`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled),
-[`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled).
+We also provide configurations for other integral range types:
+
+- [`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled),
+- [`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled),
+- [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled).
```
Comment (jlowe) [inline docs/compatibility.md:342]:
```suggestion
- [`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled)
- [`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled)
- [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled)
```

**Review by jlowe — APPROVED** [review]

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
build

**Review by jlowe — DISMISSED** [review]

File: docs/compatibility.md [hunk 21]
```diff
@@ -271,13 +271,76 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+```
+
+The results will differ between the CPU and GPU due to overflow handling.
+
+``` console
+CPU: WrappedArray([0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0])
+GPU: WrappedArray([0], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19], [19])
+```
+
+To disable byte-range windowing on the GPU, set
```
Comment (revans2) [inline docs/compatibility.md:335]: I think you mean enable not disable, either that or set the value to false.

File: docs/compatibility.md [hunk 22]
```diff
@@ -271,13 +271,76 @@ Spark stores timestamps internally relative to the JVM time zone.  Converting an
...
+[`spark.rapids.sql.window.range.byte.enabled`](configs.md#sql.window.range.byte.enabled) to true.
+
+We also provide configurations for other integral range types:
+
+- [`spark.rapids.sql.window.range.short.enabled`](configs.md#sql.window.range.short.enabled)
+- [`spark.rapids.sql.window.range.int.enabled`](configs.md#sql.window.range.int.enabled)
+- [`spark.rapids.sql.window.range.long.enabled`](configs.md#sql.window.range.short.enabled)
+
```
Comment (revans2) [inline docs/compatibility.md:343]: Can we have something explaining the defaults. It is not clear form this that ints and longs are enabled, but shorts and bytes are disabled, or why.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Arm.scala [hunk 23]
```diff
@@ -31,6 +31,15 @@ trait Arm {
     }
   }
 
+  /** Executes the provided code block and then closes the Option[resource] */
+  def withResource[T <: AutoCloseable, V](r: Option[T])(block: Option[T] => V): V = {
+    try {
+      block(r)
+    } finally {
+      r.map(_.close())
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/Arm.scala:39]: nit: technically this should be foreach instead of map. very minor.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 24]
```diff
@@ -362,6 +403,74 @@ object GpuWindowExpression {
...
+    implicit def daysToDuration(days: Long) = Duration(value, DAYS)
+    orderByType match {
+      case DType.INT8 => Scalar.fromByte(value.toByte)
+      case DType.INT16 => Scalar.fromShort(value.toShort)
+      case DType.INT32 => Scalar.fromInt(value.toInt)
+      case DType.INT64 => Scalar.fromLong(value)
+      case DType.TIMESTAMP_DAYS => Scalar.durationFromLong(DType.DURATION_DAYS, value.toDays)
+      case DType.TIMESTAMP_SECONDS => Scalar.durationFromLong(DType.DURATION_SECONDS,
+        value.toSeconds)
+      case DType.TIMESTAMP_MILLISECONDS =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:426]: I m a bit conflicted by this. Spark only supports TIMESTAMP_DAYS and TIMESTAMP_MICROSECONDS. Why are we including code for all of the others that will never be used?

Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala]: Done. removed other un-supported types

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Arm.scala (hunk 23 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/Arm.scala]: Done

File: docs/compatibility.md (hunk 22 above)
Comment (wbo4958) [reply docs/compatibility.md:343]: Done.

File: docs/compatibility.md (hunk 21 above)
Comment (wbo4958) [reply docs/compatibility.md]: Thx for the finding. Done

**Comment by wbo4958** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by mythrocks** [comment]
:+1:
